### ISSUE TITLE
All entries in Monitoring, Group3 incorrectly reporting as failed

### ISSUE THREAD

**Issue by dariushazimi** [prowler-cloud/prowler#325]
I have created all the group3 metric as per instruction in the CIS benchmark and validated, however, prowler incorrectly reports failure.

```
 3.1  [check31] Ensure a log metric filter and alarm exist for unauthorized API calls (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.2  [check32] Ensure a log metric filter and alarm exist for Management Console sign-in without MFA (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.3  [check33] Ensure a log metric filter and alarm exist for usage of root account (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.4  [check34] Ensure a log metric filter and alarm exist for IAM policy changes (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.5  [check35] Ensure a log metric filter and alarm exist for CloudTrail configuration changes (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.6  [check36] Ensure a log metric filter and alarm exist for AWS Management Console authentication failures (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.7  [check37] Ensure a log metric filter and alarm exist for disabling or scheduled deletion of customer created CMKs (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.8  [check38] Ensure a log metric filter and alarm exist for S3 bucket policy changes (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.9  [check39] Ensure a log metric filter and alarm exist for AWS Config configuration changes (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.10  [check310] Ensure a log metric filter and alarm exist for security group changes (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.11  [check311] Ensure a log metric filter and alarm exist for changes to Network Access Control Lists (NACL) (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.12  [check312] Ensure a log metric filter and alarm exist for changes to network gateways (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.13  [check313] Ensure a log metric filter and alarm exist for route table changes (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 

 3.14  [check314] Ensure a log metric filter and alarm exist for VPC changes (Scored)
       FAIL! No CloudWatch group found for CloudTrail events 
```

**Comment by toniblyx** [comment]
Hi @dariushazimi thanks for reporting this issue. Group 3 of checks is probably one of the more complex to achieve since not everybody uses the same names, filters, metrics, etc. Can you share the code you used to implement those controls? If you have some CloudFormation or scripts to do so. Did you use something like this piece of automation? https://asecure.cloud/p/monitoring_cis_benchmark/ or like https://github.com/nozaq/terraform-aws-secure-baseline Thanks!

**Comment by dariushazimi** [comment]
Hi @toniblyx 

I am using the standard published controls by AWS and here is the list:
Whithin the following page I use the controls with `"filter patterns" `
https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards.html#securityhub-standards-enabling

Next I use use the following:
https://aws-quickstart.s3.amazonaws.com/quickstart-compliance-cis-benchmark/doc/cis-benchmark-on-the-aws-cloud.pdf

There is a youtube video which explains the deployment using Yaml:
https://www.youtube.com/watch?v=01VFDHmeaDs&feature=youtu.be

and here is the yaml file
![image](https://user-images.githubusercontent.com/622048/57460162-9b466280-7242-11e9-86f2-447393d7800d.png)

Next when I enable all the controls above and run the prowler, the report shows controls are missing even though they have all been deployed.

![image](https://user-images.githubusercontent.com/622048/57460310-d9dc1d00-7242-11e9-93d8-5769871b9256.png)

**Comment by toniblyx** [comment]
Hi @dariushazimi, I have deployed the same official CFN template and ran Prowler `group3` to see it by myself and it only detects `check32` and `check33`. I'm going to work on debugging it and let you know.
![Screenshot 2019-05-14 21 27 38](https://user-images.githubusercontent.com/3985464/57742468-3dc17400-768f-11e9-94a4-8b3cd44037da.png)

**Comment by toniblyx** [comment]
hey @dariushazimi I have changed a bit the logic of `check31` to see if makes sense to be more specific for the official hardening CFN template. Can you check this code and let me know what do you think? I can adapt the rest of the checks accordingly:

```
#!/usr/bin/env bash

# Prowler - the handy cloud security tool (c) by Toni de la Fuente
#
# This Prowler check is licensed under a
# Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License.
#
# You should have received a copy of the license along with this
# work. If not, see <http://creativecommons.org/licenses/by-nc-sa/4.0/>.

CHECK_ID_check31="3.1,3.01"
CHECK_TITLE_check31="[check31] Ensure a log metric filter and alarm exist for unauthorized API calls (Scored)"
CHECK_SCORED_check31="SCORED"
CHECK_TYPE_check31="LEVEL1"
CHECK_ALTERNATE_check301="check31"

check31(){
  # "Ensure a log metric filter and alarm exist for unauthorized API calls (Scored)"
  CLOUDWATCH_GROUP=$($AWSCLI cloudtrail describe-trails $PROFILE_OPT --region $REGION --query 'trailList[*].CloudWatchLogsLogGroupArn' --output text| tr '	' '
' | awk -F: '{ print $7 }')
  if [[ $CLOUDWATCH_GROUP ]];then
    for group in $CLOUDWATCH_GROUP; do
      CLOUDWATCH_LOGGROUP_REGION=$($AWSCLI cloudtrail describe-trails $PROFILE_OPT --region $REGION --query 'trailList[*].CloudWatchLogsLogGroupArn' --output text | tr '	' '
' | grep $group | awk -F: '{ print $4 }'  | head -n 1)
      METRICFILTER_SET=$($AWSCLI logs describe-metric-filters --log-group-name $group $PROFILE_OPT --region $CLOUDWATCH_LOGGROUP_REGION --output text | awk '/UnauthorizedApiCalls/{n=NR+0} n>=NR'|awk '{ print $3 }')
      if [[ $METRICFILTER_SET ]];then
        # check if that metric filter is set correctly with /UnauthorizedOperation/ || /AccessDenied/
        METRICFILTER_FILTER=$($AWSCLI logs describe-metric-filters --log-group-name $group $PROFILE_OPT --region $CLOUDWATCH_LOGGROUP_REGION --query 'metricFilters[?filterName==`'$METRICFILTER_SET'`].[filterName,filterPattern]' --output text | awk 'BEGIN {IGNORECASE=1}; /UnauthorizedOperation/ || /AccessDenied/ {print $3};')
        for metric in $METRICFILTER_SET; do
          metric_name=$($AWSCLI logs describe-metric-filters $PROFILE_OPT --region $CLOUDWATCH_LOGGROUP_REGION --log-group-name $group --filter-name-prefix $metric --output text --query 'metricFilters[0].metricTransformations[0].metricName')
          HAS_ALARM_ASSOCIATED=$($AWSCLI cloudwatch describe-alarms $PROFILE_OPT --region $CLOUDWATCH_LOGGROUP_REGION --query 'MetricAlarms[?MetricName==`'$metric_name'`]' --output text)
          if [[ $HAS_ALARM_ASSOCIATED ]];then
            CHECK31OK="$CHECK31OK $group:$metric"
          else
            CHECK31WARN="$CHECK31WARN $group:$metric"
          fi
        done
      else
        CHECK31WARN="$CHECK31WARN $group"
      fi
    done

    if [[ $CHECK31OK ]]; then
      for group in $CHECK31OK; do
        metric=${group#*:}
        group=${group%:*}
        textPass "CloudWatch group $group found with metric filter $metric and alarms set for Unauthorized Operation and Access Denied"
      done
    fi
    if [[ $CHECK31WARN ]]; then
      for group in $CHECK31WARN; do
        case $group in
           *:*) metric=${group#*:}
                group=${group%:*}
                textFail "CloudWatch group $group found with metric filter $metric but no alarms associated"
                ;;
             *) textFail "CloudWatch group $group found but no metric filters or alarms associated"
        esac
      done
    fi
  else
    textFail "No CloudWatch group found for CloudTrail events"
  fi
}

```

**Comment by TimUbbink** [comment]
Hi @toniblyx, 

Sorry if I necro anything, but we are encountering the same group3 issues on our side. We are getting, what I think are, false positives on 32, 35, 36, 37, and 312. The only thing is that they are not as consistent as Darius reported since the rest of the group 3 checks pass. The error states that the Cloudwatch group is found with metric filter but no alarms associated.

I've checked the console, CLI, and code but everything should be configured properly. I also adapted https://asecure.cloud/p/monitoring_cis_benchmark/ with our template, but was unsuccessful to let these checks pass. 

If you need any more information, please let me know.

**Comment by toniblyx** [comment]
This should be fixed as soon as I merge PR #366

**Comment by toniblyx** [comment]
this should be fixed already in PR #366. Please, check it out with the latest code in master branch and re open it with details if needed. Thanks
